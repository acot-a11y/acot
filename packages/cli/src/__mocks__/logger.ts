import { MockLogger } from '@acot/mock';

export const createMockLogger = (): MockLogger => new MockLogger();
