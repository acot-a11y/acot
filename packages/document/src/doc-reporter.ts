import ms from 'pretty-ms';
import { pickup } from '@acot/html-pickup';
import type { RuleId, TestcaseResult } from '@acot/types';
import chalk from 'chalk';
import indent from 'indent-string';
import logSymbols from 'log-symbols';
import plur from 'plur';
import stripAnsi from 'strip-ansi';
import { generateDocPath } from './doc-code';
import type { DocError } from './doc-error';
import type { DocResult } from './doc-result';

const symbols = {
  success: stripAnsi(logSymbols.success),
  warning: stripAnsi(logSymbols.warning),
  error: stripAnsi(logSymbols.error),
};

export type DocReporterConfig = {
  origin: string;
  color: boolean;
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
};

export class DocReporter {
  private _stdout: NodeJS.WritableStream;
  private _config: DocReporterConfig;
  private _chalk: chalk.Chalk;

  public constructor(config: Partial<DocReporterConfig> = {}) {
    this._config = {
      origin: 'http://localhost:1234',
      color: true,
      stdout: process.stdout,
      stderr: process.stderr,
      ...config,
    };

    this._stdout = this._config.stdout;

    this._chalk = new chalk.Instance({
      level: this._config.color ? 1 : 0,
    });
  }

  public async onTestcaseComplete(
    path: string,
    _id: RuleId,
    results: TestcaseResult[],
  ): Promise<void> {
    const chk = this._chalk;

    const data = results.reduce(
      (acc, cur) => {
        acc.duration += cur.duration;
        acc.pass += cur.status === 'pass' ? 1 : 0;
        acc.warning += cur.status === 'warn' ? 1 : 0;
        acc.error += cur.status === 'error' ? 1 : 0;
        return acc;
      },
      {
        duration: 0,
        pass: 0,
        warning: 0,
        error: 0,
      },
    );

    let output = `${path} `;
    output += chk` {gray ${symbols.success}} {bold.gray ${data.pass}}`;
    output += chk` {gray ${symbols.warning}} {bold.gray ${data.warning}}`;
    output += chk` {gray ${symbols.error}} {bold.gray ${data.error}}`;
    output += chk`  {gray (${ms(data.duration)})}`;

    this._stdout.write(`${output}\n`);
  }

  public async onComplete(result: DocResult): Promise<void> {
    const { origin } = this._config;
    const chk = this._chalk;

    let output = '\n';

    const groups = result.errors.reduce<Map<string, DocError[]>>((acc, cur) => {
      if (acc.has(cur.code.rule)) {
        acc.get(cur.code.rule)!.push(cur);
      } else {
        acc.set(cur.code.rule, [cur]);
      }
      return acc;
    }, new Map());

    groups.forEach((local, rule) => {
      const fail = chk.bold.red.inverse(' FAIL ');
      const title = chk.bold.underline(rule);
      const stat = chk.gray`(${local.length} ${plur('error', local.length)})`;

      output += `${fail} ${title} ${stat}\n\n`;

      local.forEach((error) => {
        output += chk`{red ${logSymbols.error}} {bold.red ${error.message}}`;
        output += '\n';

        error.results.forEach((result) => {
          output += indent(chk`{bold ${result.message}}\n`, 4);

          const meta: string[] = [];

          if (result.selector != null) {
            const html = pickup(error.code.html, result.selector, {
              color: false,
            });

            if (html != null) {
              meta.push(html);
              meta.push(`at "${result.selector}"`);
            }
          }

          meta.push(origin + generateDocPath(error.code));

          const metaStr = meta
            .map((m, i) => {
              return i === meta.length - 1 ? `└─ ${m}` : `├─ ${m}`;
            })
            .join('\n');

          output += indent(chk.gray(metaStr), 4);
          output += '\n';
        });
      });

      output += '\n';
    });

    const passed = Array.from(new Set(result.passes.map((code) => code.rule)));
    const failed = Array.from(groups.keys());
    const total = Array.from(new Set([...passed, ...failed]));

    const all = {
      total: total.length,
      passed: passed.length,
      failed: failed.length,
    };

    const cases = {
      total: result.passes.length + result.errors.length,
      passed: result.passes.length,
      failed: result.errors.length,
    };

    const display = (obj: {
      total: number;
      passed: number;
      failed: number;
    }) => {
      const items = [];

      if (obj.passed > 0) {
        items.push(chk.bold.green(`${obj.passed} passed`));
      }

      if (obj.failed > 0) {
        items.push(chk.bold.red(`${obj.failed} failed`));
      }

      items.push(`${obj.total} total`);

      return items.join(', ');
    };

    const footer: string[] = [];
    footer.push(chk`{bold Rules:     } ${display(all)}`);
    footer.push(chk`{bold Test Cases:} ${display(cases)}`);

    output += footer.join('\n');

    this._stdout.write(chk.reset(`${output}\n\n`));
  }
}
