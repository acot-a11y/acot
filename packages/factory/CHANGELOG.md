# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.19](https://github.com/acot-a11y/acot/compare/v0.0.18...v0.0.19) (2022-06-05)

**Note:** Version bump only for package @acot/factory

## [0.0.18](https://github.com/acot-a11y/acot/compare/v0.0.17...v0.0.18) (2022-05-11)

**Note:** Version bump only for package @acot/factory

## [0.0.17](https://github.com/acot-a11y/acot/compare/v0.0.17-canary.3...v0.0.17) (2022-05-10)

**Note:** Version bump only for package @acot/factory

## [0.0.17-canary.0](https://github.com/acot-a11y/acot/compare/v0.0.16...v0.0.17-canary.0) (2022-05-05)

**Note:** Version bump only for package @acot/factory

## [0.0.16](https://github.com/acot-a11y/acot/compare/v0.0.15...v0.0.16) (2022-05-01)

**Note:** Version bump only for package @acot/factory

## [0.0.15](https://github.com/acot-a11y/acot/compare/v0.0.14...v0.0.15) (2022-03-27)

**Note:** Version bump only for package @acot/factory

## [0.0.14](https://github.com/acot-a11y/acot/compare/v0.0.14-canary.2...v0.0.14) (2022-01-03)

**Note:** Version bump only for package @acot/factory

## [0.0.13](https://github.com/acot-a11y/acot/compare/v0.0.13-canary.0...v0.0.13) (2021-10-10)

**Note:** Version bump only for package @acot/factory

## [0.0.12](https://github.com/acot-a11y/acot/compare/v0.0.11...v0.0.12) (2021-04-18)

**Note:** Version bump only for package @acot/factory

## [0.0.11](https://github.com/acot-a11y/acot/compare/v0.0.10...v0.0.11) (2021-04-12)

**Note:** Version bump only for package @acot/factory

## [0.0.10](https://github.com/acot-a11y/acot/compare/v0.0.10-canary.0...v0.0.10) (2021-04-11)

**Note:** Version bump only for package @acot/factory

## 0.0.10-canary.0 (2021-04-04)

### Bug Fixes

- Fix omission of version specification in `debug` package ([85f8ea4](https://github.com/acot-a11y/acot/commit/85f8ea44c7b029301dbcd6bceef427fda35972b6))

### Features

- Add rule result for pretty reporter ([f668f8a](https://github.com/acot-a11y/acot/commit/f668f8a56e730ea31ece53f23d7fd2629e456211))
- Initial implements ([68f46a2](https://github.com/acot-a11y/acot/commit/68f46a250de7793795678ece40d23d927ddd075c))
- **core:** switch to `help` from `tags` ([7ac5678](https://github.com/acot-a11y/acot/commit/7ac56780e46c26f94aeda3f0c1d1482f47521acb))

## [0.0.8-canary.0](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.7...@acot/factory@0.0.8-canary.0) (2021-03-30)

**Note:** Version bump only for package @acot/factory

## [0.0.7](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.6...@acot/factory@0.0.7) (2021-01-30)

### Features

- **core:** switch to `help` from `tags` ([7ac5678](https://github.com/acot-a11y/acot/commit/7ac56780e46c26f94aeda3f0c1d1482f47521acb))

## [0.0.6](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.6-canary.0...@acot/factory@0.0.6) (2021-01-13)

**Note:** Version bump only for package @acot/factory

## [0.0.6-canary.0](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.5...@acot/factory@0.0.6-canary.0) (2021-01-13)

**Note:** Version bump only for package @acot/factory

## [0.0.5](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.4...@acot/factory@0.0.5) (2021-01-01)

**Note:** Version bump only for package @acot/factory

## [0.0.4](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.4-canary.0...@acot/factory@0.0.4) (2020-12-29)

**Note:** Version bump only for package @acot/factory

## [0.0.4-canary.0](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.3...@acot/factory@0.0.4-canary.0) (2020-12-29)

**Note:** Version bump only for package @acot/factory

## [0.0.3](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.3-canary.2...@acot/factory@0.0.3) (2020-12-22)

**Note:** Version bump only for package @acot/factory

## [0.0.3-canary.2](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.3-canary.1...@acot/factory@0.0.3-canary.2) (2020-12-19)

### Features

- Add rule result for pretty reporter ([f668f8a](https://github.com/acot-a11y/acot/commit/f668f8a56e730ea31ece53f23d7fd2629e456211))

## [0.0.3-canary.1](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.3-canary.0...@acot/factory@0.0.3-canary.1) (2020-12-16)

**Note:** Version bump only for package @acot/factory

## [0.0.3-canary.0](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.2...@acot/factory@0.0.3-canary.0) (2020-12-14)

### Bug Fixes

- Fix omission of version specification in `debug` package ([85f8ea4](https://github.com/acot-a11y/acot/commit/85f8ea44c7b029301dbcd6bceef427fda35972b6))

## [0.0.2](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.2-canary.3...@acot/factory@0.0.2) (2020-12-10)

**Note:** Version bump only for package @acot/factory

## [0.0.2-canary.3](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.2-canary.2...@acot/factory@0.0.2-canary.3) (2020-11-23)

**Note:** Version bump only for package @acot/factory

## [0.0.2-canary.2](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.2-canary.1...@acot/factory@0.0.2-canary.2) (2020-11-08)

**Note:** Version bump only for package @acot/factory

## [0.0.2-canary.1](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.2-canary.0...@acot/factory@0.0.2-canary.1) (2020-11-08)

**Note:** Version bump only for package @acot/factory

## [0.0.2-canary.0](https://github.com/acot-a11y/acot/compare/@acot/factory@0.0.1...@acot/factory@0.0.2-canary.0) (2020-11-02)

**Note:** Version bump only for package @acot/factory

## 0.0.1 (2020-10-31)

### Features

- Initial implements ([68f46a2](https://github.com/acot-a11y/acot/commit/68f46a250de7793795678ece40d23d927ddd075c))
