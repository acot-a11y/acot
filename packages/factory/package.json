{
  "name": "@acot/factory",
  "version": "0.0.6-canary.0",
  "description": "A factory toolset for various acot objects.",
  "homepage": "https://github.com/acot-a11y/acot/tree/main/packages/factory",
  "bugs": {
    "url": "https://github.com/acot-a11y/acot/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/acot-a11y/acot.git",
    "directory": "packages/factory"
  },
  "license": "MIT",
  "author": "wadackel <wadackel@gmail.com>",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!__mocks__",
    "!__tests__"
  ],
  "scripts": {
    "build": "tsc"
  },
  "dependencies": {
    "@acot/types": "0.0.6-canary.0",
    "deepmerge": "^4.2.2",
    "is-plain-object": "^5.0.0",
    "type-fest": "^0.20.2"
  },
  "publishConfig": {
    "access": "public"
  }
}
