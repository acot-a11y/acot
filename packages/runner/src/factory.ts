import type { RunnerOptions, RunnerFactory } from '@acot/types';

export const createRunnerFactory = <T extends RunnerOptions = {}>(
  descriptor: RunnerFactory<T>,
): RunnerFactory<T> => descriptor;
