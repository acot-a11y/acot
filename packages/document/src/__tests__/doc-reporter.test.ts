import { PassThrough } from 'stream';
import { createTestcaseResult } from '@acot/factory';
import stripAnsi from 'strip-ansi';
import { DocReporter } from '../doc-reporter';
import { createDocCode } from '../__mocks__/doc-code';
import { createDocError } from '../__mocks__/doc-error';
import { createDocResult } from '../__mocks__/doc-result';

const createMockStream = (data: string[]) => {
  const stream = new PassThrough();

  stream.on('data', (buf) => {
    data.push(buf.toString().replace(/\n$/, ''));
  });

  return stream;
};

describe('DocReporter', () => {
  const html = `
<div class="wrapper">
  <section>
    <div id="element">
      <p>Content</p>
      <p>Content</p>
      <p>Content</p>
    </div>

    <img src="path/to.png" alt="Alternative content">
  </section>
</div>
    `.trim();

  let reporter: DocReporter;
  let output: {
    stdout: string[];
    stderr: string[];
  };

  beforeEach(() => {
    output = {
      stdout: [],
      stderr: [],
    };

    reporter = new DocReporter({
      stdout: createMockStream(output.stdout),
      stderr: createMockStream(output.stderr),
    });
  });

  test('empty', async () => {
    await reporter.onComplete(createDocResult());

    const stdout = output.stdout.join('');

    expect(stdout).not.toBe(stripAnsi(stdout));
    expect(stripAnsi(stdout)).toMatchSnapshot();
  });

  test('color disable', async () => {
    const instance = new DocReporter({
      color: false,
      stdout: createMockStream(output.stdout),
      stderr: createMockStream(output.stderr),
    });

    await instance.onComplete(createDocResult());

    const stdout = output.stdout.join('');

    expect(stdout).toBe(stripAnsi(stdout));
  });

  test('passed and errors', async () => {
    await reporter.onComplete(
      createDocResult({
        skips: [createDocCode({ rule: 'skip-name1', id: '1' })],
        passes: [
          createDocCode({ rule: 'pass-name1', id: '1' }),
          createDocCode({ rule: 'pass-name1', id: '2' }),
        ],
        errors: [
          createDocError({
            code: createDocCode({
              rule: 'error-name1',
              id: '1',
              html,
            }),
            message: 'Error message',
            results: [
              createTestcaseResult({
                status: 'error',
                selector: '#element',
                message: 'Element error message.',
              }),
              createTestcaseResult({
                status: 'error',
                message: 'Global error message.',
              }),
            ],
          }),
          createDocError({
            code: createDocCode({
              rule: 'error-name2',
              id: '2',
              html,
            }),
            message: 'Error message',
            results: [
              createTestcaseResult({
                status: 'error',
                message: 'Element error message.',
                selector: 'img',
              }),
            ],
          }),
          createDocError({
            code: createDocCode({
              rule: 'error-name3',
              id: '3',
              html,
            }),
            message: 'Error message',
            results: [],
          }),
          createDocError({
            code: createDocCode({
              rule: 'error-name4',
              id: '4',
              html,
            }),
            message: 'Error message',
            results: [
              createTestcaseResult({
                status: 'error',
                selector: '#element',
                message: 'Error!!',
              }),
            ],
          }),
        ],
      }),
    );

    const stdout = output.stdout.join('');

    expect(stdout).not.toBe(stripAnsi(stdout));
    expect(stripAnsi(stdout)).toMatchSnapshot();
  });

  test('only passed', async () => {
    await reporter.onComplete(
      createDocResult({
        passes: [
          createDocCode({ rule: 'pass-name1', id: '1' }),
          createDocCode({ rule: 'pass-name2', id: '2' }),
        ],
      }),
    );

    const stdout = output.stdout.join('');

    expect(stdout).not.toBe(stripAnsi(stdout));
    expect(stripAnsi(stdout)).toMatchSnapshot();
  });

  test('only errors', async () => {
    await reporter.onComplete(
      createDocResult({
        errors: [
          createDocError({
            code: createDocCode({
              rule: 'error-name1',
              id: '1',
              html,
            }),
            message: 'Error message',
            results: [
              createTestcaseResult({
                status: 'error',
                selector: '#element',
                message: 'Element error message.',
              }),
            ],
          }),
        ],
      }),
    );

    const stdout = output.stdout.join('');

    expect(stdout).not.toBe(stripAnsi(stdout));
    expect(stripAnsi(stdout)).toMatchSnapshot();
  });
});
