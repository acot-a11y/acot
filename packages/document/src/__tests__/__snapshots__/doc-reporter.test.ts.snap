// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocReporter empty 1`] = `
"
Rules:      0 total
Test Cases: 0 total

"
`;

exports[`DocReporter only errors 1`] = `
"
 FAIL  error-name1 (1 error)

✖ Error message
    Element error message.
    ├─ <div id=\\"element\\"><p>Content</p><p>Content</p><p>Content</p></div>
    ├─ at \\"#element\\"
    └─ http://localhost:1234/error-name1/1

Rules:      1 failed, 1 total
Test Cases: 1 failed, 1 total

"
`;

exports[`DocReporter only passed 1`] = `
"
Rules:      2 passed, 2 total
Test Cases: 2 passed, 2 total

"
`;

exports[`DocReporter passed and errors 1`] = `
"
 FAIL  error-name1 (1 error)

✖ Error message
    Element error message.
    ├─ <div id=\\"element\\"><p>Content</p><p>Content</p><p>Content</p></div>
    ├─ at \\"#element\\"
    └─ http://localhost:1234/error-name1/1
    Global error message.
    └─ http://localhost:1234/error-name1/1

 FAIL  error-name2 (1 error)

✖ Error message
    Element error message.
    ├─ <img src=\\"path/to.png\\" alt=\\"Alternative content\\" />
    ├─ at \\"img\\"
    └─ http://localhost:1234/error-name2/2

 FAIL  error-name3 (1 error)

✖ Error message

 FAIL  error-name4 (1 error)

✖ Error message
    Error!!
    ├─ <div id=\\"element\\"><p>Content</p><p>Content</p><p>Content</p></div>
    ├─ at \\"#element\\"
    └─ http://localhost:1234/error-name4/4

Rules:      1 passed, 4 failed, 5 total
Test Cases: 2 passed, 4 failed, 6 total

"
`;
